### PR TITLE
binance: add futures exchange api queries

### PR DESCRIPTION
**Pull request by austin362667**
1. add futures exchange api queries.
2. TODO:  fix query trades, closed orders futures symbol not found
3. TODO: query account with spot, margin, futures.
4. TODO: output type of toGlobalMarkets is the same as toGlobalFuturesMarkets

### CODE REVIEW

File: pkg/exchange/binance/exchange.go [hunk 1]
```diff
@@ -169,6 +178,22 @@ func (e *Exchange) QueryTickers(ctx context.Context, symbol ...string) (map[stri
...
+
+	if e.IsFutures {
+		exchangeInfo, err := e.futuresClient.NewExchangeInfoService().Do(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		markets := types.MarketMap{}
+		for _, symbol := range exchangeInfo.Symbols {
+			log.Info(symbol)
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:190]: remove debug log

File: pkg/exchange/binance/exchange.go [hunk 2]
```diff
@@ -431,15 +456,61 @@ func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
...
 	return a, nil
 }
 
+func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
+	futuresAccount, err := e.QueryFuturesAccount(ctx)
+	if !(err != nil) {
+		return futuresAccount, nil
+	}
+
+	account, err := e.QuerySpotAccount(ctx)
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:506]: you can directly return the value

File: pkg/exchange/binance/exchange.go [hunk 3]
```diff
@@ -510,7 +581,9 @@ func (e *Exchange) QueryClosedOrders(ctx context.Context, symbol string, since,
 
 		binanceOrders, err := req.Do(ctx)
 		if err != nil {
-			return orders, err
+			if !strings.Contains(err.Error(), "-1121") {
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:584]: what's the -1121 error code? why not just return the error?

File: pkg/exchange/binance/exchange.go [hunk 4]
```diff
@@ -1017,28 +1090,44 @@ func (e *Exchange) QueryTrades(ctx context.Context, symbol string, options *type
...
 			req.Limit(1000)
 		}
 
+		//if options.StartTime != nil {
+		//	req.StartTime(options.StartTime.UnixNano() / int64(time.Millisecond))
+		//}
+		//
+		//if options.EndTime != nil {
+		//	req.EndTime(options.EndTime.UnixNano() / int64(time.Millisecond))
+		//}
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:1107]: if unused please remove it

File: pkg/exchange/binance/exchange.go (hunk 3 above)
Comment (austin362667) [reply pkg/exchange/binance/exchange.go:584]: -1121 means symbol not found.

File: pkg/exchange/binance/exchange.go (hunk 1 above)
Comment (austin362667) [reply pkg/exchange/binance/exchange.go]: done

File: pkg/exchange/binance/exchange.go (hunk 2 above)
Comment (austin362667) [reply pkg/exchange/binance/exchange.go]: ok, thx!

File: pkg/exchange/binance/exchange.go (hunk 3 above)
Comment (austin362667) [reply pkg/exchange/binance/exchange.go]: fixed, thanks a lot

File: pkg/exchange/binance/exchange.go (hunk 4 above)
Comment (austin362667) [reply pkg/exchange/binance/exchange.go]: removed

File: pkg/exchange/binance/exchange.go [hunk 5]
```diff
@@ -431,15 +446,56 @@ func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
...
+		CanTrade:    account.CanTrade,                    // if can trade
+		CanWithdraw: account.CanWithdraw,                 // if can transfer out asset
 	}
 	a.UpdateBalances(balances)
 	return a, nil
 }
 
+func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
+	futuresAccount, err := e.QueryFuturesAccount(ctx)
+	if !(err != nil) {
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:492]: ??

File: pkg/exchange/binance/exchange.go [hunk 6]
```diff
@@ -431,15 +446,56 @@ func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
...
+		CanDeposit:  account.CanDeposit,                  // if can transfer in asset
+		CanTrade:    account.CanTrade,                    // if can trade
+		CanWithdraw: account.CanWithdraw,                 // if can transfer out asset
 	}
 	a.UpdateBalances(balances)
 	return a, nil
 }
 
+func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
+	futuresAccount, err := e.QueryFuturesAccount(ctx)
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:491]: did not check IsFutures flag?

File: pkg/exchange/binance/exchange.go [hunk 7]
```diff
@@ -1017,9 +1072,8 @@ func (e *Exchange) QueryTrades(ctx context.Context, symbol string, options *type
 		return trades, nil
 	} else if e.IsFutures {
 		var remoteTrades []*futures.AccountTrade
-		req := e.futuresClient.NewListAccountTradeService(). // IsIsolated(e.IsIsolatedFutures).
+		req := e.futuresClient.NewListAccountTradeService(). //IsIsolated(e.IsIsolatedFutures).
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:1075]: IsIsolated is not used?

File: pkg/service/sync.go [hunk 8]
```diff
@@ -25,17 +25,20 @@ type SyncService struct {
...
 		log.Infof("syncing %s %s trades...", exchange.Name(), symbol)
-		if err := s.TradeService.Sync(ctx, exchange, symbol, startTime); err != nil {
-			return err
-		}
-
-		log.Infof("syncing %s %s orders...", exchange.Name(), symbol)
-		if err := s.OrderService.Sync(ctx, exchange, symbol, startTime); err != nil {
-			return err
+		markets, err := exchange.QueryMarkets(ctx)
+		if err != nil {
```
Comment (c9s) [inline pkg/service/sync.go:29]: err != nil ?

File: pkg/service/sync.go [hunk 9]
```diff
@@ -25,17 +25,20 @@ type SyncService struct {
...
 	for _, symbol := range symbols {
 		log.Infof("syncing %s %s trades...", exchange.Name(), symbol)
-		if err := s.TradeService.Sync(ctx, exchange, symbol, startTime); err != nil {
-			return err
-		}
-
-		log.Infof("syncing %s %s orders...", exchange.Name(), symbol)
-		if err := s.OrderService.Sync(ctx, exchange, symbol, startTime); err != nil {
-			return err
+		markets, err := exchange.QueryMarkets(ctx)
```
Comment (c9s) [inline pkg/service/sync.go:28]: you should use load markets with cache here, there is a function.

Comment (c9s) [reply pkg/service/sync.go:28]: check cache.go

File: pkg/types/account.go [hunk 10]
```diff
@@ -227,18 +228,18 @@ type Account struct {
...
-	TotalInitialMargin          fixedpoint.Value           `json:"totalInitialMargin"`
-	TotalMaintMargin            fixedpoint.Value           `json:"totalMaintMargin"`
-	TotalMarginBalance          fixedpoint.Value           `json:"totalMarginBalance"`
-	TotalOpenOrderInitialMargin fixedpoint.Value           `json:"totalOpenOrderInitialMargin"`
-	TotalPositionInitialMargin  fixedpoint.Value           `json:"totalPositionInitialMargin"`
-	TotalUnrealizedProfit       fixedpoint.Value           `json:"totalUnrealizedProfit"`
-	TotalWalletBalance          fixedpoint.Value           `json:"totalWalletBalance"`
-	UpdateTime                  int64                      `json:"updateTime"`
+	Assets                      FuturesAssetMap    `json:"assets"`
+	FeeTier                     int                `json:"feeTier"`
```
Comment (c9s) [inline pkg/types/account.go:232]: This fee tier might not be shared between different exchange

File: pkg/types/account.go [hunk 11]
```diff
@@ -227,18 +228,18 @@ type Account struct {
 
 type FuturesAccountInfo struct {
 	// Futures fields
-	Assets                      map[Asset]FuturesUserAsset `json:"assets"`
-	FeeTier                     int                        `json:"feeTier"`
-	MaxWithdrawAmount           fixedpoint.Value           `json:"maxWithdrawAmount"`
```
Comment (c9s) [inline pkg/types/account.go:232]: MaxWithdrawAmount is not used currently, let's try to narrow down the fields.

File: pkg/exchange/binance/exchange.go [hunk 12]
```diff
@@ -431,15 +446,56 @@ func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
...
+	accountBalances, err := e.futuresClient.NewGetBalanceService().Do(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var balances = map[string]types.Balance{}
+	for _, b := range accountBalances {
+		balances[b.Asset] = types.Balance{
+			Currency:  b.Asset,
+			Available: fixedpoint.Must(fixedpoint.NewFromString(b.AvailableBalance)),
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:475]: avoid using Must here, we should check error

File: pkg/exchange/binance/exchange.go [hunk 13]
```diff
@@ -426,20 +451,70 @@ func (e *Exchange) QueryAccount(ctx context.Context) (*types.Account, error) {
...
+	if e.IsFutures {
+		account, err = e.QueryFuturesAccount(ctx)
+	} else if e.IsIsolatedMargin {
+		account, err = e.QueryIsolatedMarginAccount(ctx)
+	} else if e.IsMargin {
+		account, err = e.QueryMarginAccount(ctx)
+	} else {
+		account, err = e.QuerySpotAccount(ctx)
+	}
+	if err != nil {
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:512]: you can just return ? `return account, err`

File: pkg/service/sync.go [hunk 14]
```diff
@@ -25,17 +25,24 @@ type SyncService struct {
 func (s *SyncService) SyncSessionSymbols(ctx context.Context, exchange types.Exchange, startTime time.Time, symbols ...string) error {
 	for _, symbol := range symbols {
 		log.Infof("syncing %s %s trades...", exchange.Name(), symbol)
-		if err := s.TradeService.Sync(ctx, exchange, symbol, startTime); err != nil {
+		// TODO: bbgo import cycle error
+		//markets, err := bbgo.LoadExchangeMarketsWithCache(ctx, exchange)
```
Comment (c9s) [inline pkg/service/sync.go:29]: you can pull out this `LoadExchangeMarketsWithCache` to pkg/cache

File: pkg/strategy/funding/strategy.go [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+			if kline.Interval != detection.Interval {
+				continue
+			}
+
+			closePriceF := kline.GetClose()
+			closePrice := fixedpoint.NewFromFloat(closePriceF)
+
+			var ma types.Float64Indicator
+
+			switch strings.ToLower(detection.MovingAverageType) {
```
Comment (c9s) [inline pkg/strategy/funding/strategy.go:182]: you can pull out this section to the outside of this closure? make this closure smaller

File: pkg/strategy/funding/strategy.go [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+			var lastMA = ma.Last()
+
+			// skip if the closed price is under the moving average
+			if closePrice.Float64() < lastMA {
+				log.Infof("skip %s support closed price %f > last ma %f", s.Symbol, closePrice.Float64(), lastMA)
+				return
+			}
+
+			premiumIndex, err := binanceExchange.QueryPremiumIndex(ctx, s.Symbol)
```
Comment (c9s) [inline pkg/strategy/funding/strategy.go:208]: add QueryPremiumIndex method to the FuturesExchange interface, you can assert the exchange from the beginning, and check the error outside this closure

File: pkg/strategy/funding/strategy.go [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+			fundingRate := premiumIndex.LastFundingRate
+
+			if fundingRate >= s.FundingRate.High {
+				s.Notifiability.Notify("%s funding rate %s is too high! threshold %s",
+					s.Symbol,
+					fundingRate.Percentage(),
+					s.FundingRate.High.Percentage(),
+				)
+			} else {
+				return
```
Comment (c9s) [inline pkg/strategy/funding/strategy.go:222]: add a log printing why did we skip?

File: pkg/strategy/funding/strategy.go [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+				)
+				s.Notifiability.Notify(kline)
+
+				baseBalance, ok := session.Account.Balance(s.Market.BaseCurrency)
+				if !ok {
+					return
+				}
+
+				if baseBalance.Available > 0 && baseBalance.Total() < s.MaxPosition {
+					log.Infof("start to short position, selling futures..")
```
Comment (c9s) [inline pkg/strategy/funding/strategy.go:243]: `opening a short position`

File: pkg/exchange/binance/exchange.go [hunk 19]
```diff
@@ -1203,6 +1276,18 @@ func (e *Exchange) QueryFundingRateHistory(ctx context.Context, symbol string) (
 	}, nil
 }
 
+func (e *Exchange) QueryPositionRisk(ctx context.Context, symbol string) (*types.PositionRisk, error) {
```
Comment (c9s) [inline pkg/exchange/binance/exchange.go:1279]: add this method to the FuturesExchange interface?

File: pkg/service/sync.go (hunk 14 above)
Comment (austin362667) [reply pkg/service/sync.go:29]: pkg/cache or pkg/service/cache ?

File: pkg/strategy/funding/strategy.go [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+		Interval types.Interval `json:"interval"`
+
+		// MovingAverageType is the moving average indicator type that we want to use,
+		// it could be SMA or EWMA
+		MovingAverageType string `json:"movingAverageType"`
+
+		// MovingAverageInterval is the interval of k-lines for the moving average indicator to calculate,
+		// it could be "1m", "5m", "1h" and so on.  note that, the moving averages are calculated from
+		// the k-line data we subscribed
+		MovingAverageInterval types.Interval `json:"movingAverageInterval"`
```
Comment (c9s) [inline pkg/strategy/funding/strategy.go:55]: There is a struct combining both Interval and Window,

take a look at `types.IntervalWindow` or `bbgo.MovingAverageSettings`, see if we can reuse them and make the code clean

File: pkg/service/sync.go (hunk 14 above)
Comment (austin362667) [reply pkg/service/sync.go:29]: refactored move cache.go & home.go to pkg/cache

**Review by c9s — APPROVED** [review]

**Pull request merged**